v23/context/context_test.go

```diff
@@ -320,6 +322,27 @@ func TestRootCancelChain(t *testing.T) {
 	}
 }
 
+func TestRootCancelGoroutineLeak(t *testing.T) {
+	rootCtx, rootcancel := context.RootContext()
+	const iterations = 1000
+	for i := 0; i != iterations; i++ {
+		_, cancel := context.WithRootCancel(rootCtx)
+		cancel()
+	}
+	buf := make([]byte, 2<<20)
+	buf = buf[:runtime.Stack(buf, true)]
+	count := 0
+	for _, g := range strings.Split(string(buf), "\n\n") {
```

Comment:
this is likely going to be flaky, at least in terms of getting to 0. You may want to just check that some, say a third, of the goroutines have exited since there's no way any of them would exit without your change for this test.

Reply:
Great point. This is indeed flaky after running `go test -count=1000 -run TestRootCancelGoroutineLeak`.
There were a few options I thought of:
1. Do what you suggested, which is check some percentage of them exited.
2. Make the test sleep for say 5-10ms after canceling all of the contexts and then check.
3. Adjust the context API to expose whether the context's goroutine has ended via some internal context value check (e.g. similar to checking whether the context is done).
4. Move the test to be an internal test (i.e. white box test) and adjust the test to wait for all of the contexts' goroutines to have finished via some internal cancel context value check (also similar to checking whether the context is done).

Approach 1 has been a little flaky in terms of figuring out the threshold.

Approach 2 seems to work. From what I've tested, it works for 10000 runs of the tests, but there's still always the possibility that it doesn't quite work depending on how go schedules the goroutines. It's also not great to have a test sleep and arbitrarily take up unnecessary time. This approach could  also be done in combination with approach 1 but I prefer checking that all goroutines have really exited.

Approach 3 exposes unnecessary internals of the context package for the purposes of a test which isn't great imo.

Approach 4 requires some finagling to propagate the goroutine's changed value to the cancel context which violates some boundaries (or requires some more heavy handed adjustments).

I've implemented approach 2 since it's pretty easy to do and the least intrusive to the general codebase. Let me know your thoughts and happy to switch it around as needed. The following has run without any problems from my testing so far though.
```
go test -count=10000 -run TestRootCancelGoroutineLeak$ 
```